server/src/main/java/org/elasticsearch/index/codec/vectors/es818/ES818FlatVectorsReader.java

```diff
@@ -0,0 +1,340 @@
+@SuppressForbidden(reason = "Copied from lucene")
+public class ES818FlatVectorsReader extends FlatVectorsReader {
+
+    private static final long SHALLOW_SIZE = RamUsageEstimator.shallowSizeOfInstance(ES818FlatVectorsReader.class);
+
+    private final IntObjectHashMap<FieldEntry> fields = new IntObjectHashMap<>();
+    private final IndexInput vectorData;
+    private final FieldInfos fieldInfos;
+
+    public ES818FlatVectorsReader(SegmentReadState state, FlatVectorsScorer scorer) throws IOException {
+        super(scorer);
+        int versionMeta = readMetadata(state);
+        this.fieldInfos = state.fieldInfos;
+        boolean success = false;
+        try {
+            vectorData = openDataInput(
+                state,
+                versionMeta,
+                ES818FlatVectorsFormat.VECTOR_DATA_EXTENSION,
+                ES818FlatVectorsFormat.VECTOR_DATA_CODEC_NAME,
+                // Flat formats are used to randomly access vectors from their node ID that is stored
+                // in the HNSW graph.
+                state.context.withReadAdvice(ReadAdvice.RANDOM)
+            );
+            success = true;
+        } finally {
+            if (success == false) {
+                IOUtils.closeWhileHandlingException(this);
+            }
+        }
+    }
+
+    private int readMetadata(SegmentReadState state) throws IOException {
+        String metaFileName = IndexFileNames.segmentFileName(
+            state.segmentInfo.name,
+            state.segmentSuffix,
+            ES818FlatVectorsFormat.META_EXTENSION
+        );
+        int versionMeta = -1;
+        try (ChecksumIndexInput meta = state.directory.openChecksumInput(metaFileName)) {
+            Throwable priorE = null;
+            try {
+                versionMeta = CodecUtil.checkIndexHeader(
+                    meta,
+                    ES818FlatVectorsFormat.META_CODEC_NAME,
+                    ES818FlatVectorsFormat.VERSION_START,
+                    ES818FlatVectorsFormat.VERSION_CURRENT,
+                    state.segmentInfo.getId(),
+                    state.segmentSuffix
+                );
+                readFields(meta, state.fieldInfos);
+            } catch (Throwable exception) {
+                priorE = exception;
+            } finally {
+                CodecUtil.checkFooter(meta, priorE);
+            }
+        }
+        return versionMeta;
+    }
+
+    private static IndexInput openDataInput(
+        SegmentReadState state,
+        int versionMeta,
+        String fileExtension,
+        String codecName,
+        IOContext context
+    ) throws IOException {
+        String fileName = IndexFileNames.segmentFileName(state.segmentInfo.name, state.segmentSuffix, fileExtension);
+        // use direct IO for accessing raw vector data for searches
+        IndexInput in = context.context() == IOContext.Context.DEFAULT && state.directory instanceof DirectIOIndexInputSupplier did
```

Comment:
Isn't the context always [RANDOM](https://github.com/elastic/elasticsearch/pull/125921/files#diff-49352df03f1b544985341a92d6fb5697b6ca578c1593a85ec35f44fe45d93805R76) at this point?
How do we differentiate merges from search?

Reply:
Merges have a context of `Context.MERGE`. I saw a 40% drop in force-merge performance before I put this check in here.